query/src/clusters/cluster.rs

```diff
@@ -12,140 +12,233 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::collections::hash_map::Entry::Occupied;
-use std::collections::hash_map::Entry::Vacant;
-use std::collections::HashMap;
-use std::net::IpAddr;
 use std::sync::Arc;
+use std::time::Duration;
 
+use rand::{Rng, thread_rng};
+
+use common_arrow::arrow_flight::flight_service_client::FlightServiceClient;
 use common_exception::ErrorCode;
 use common_exception::Result;
-use common_infallible::Mutex;
-use common_store_api_sdk::DNSResolver;
+use common_management::{NamespaceApi, NamespaceMgr, NodeInfo};
+use common_runtime::tokio;
+use common_runtime::tokio::sync::Mutex;
+use common_runtime::tokio::time::sleep as tokio_async_sleep;
 
-use crate::clusters::address::Address;
-use crate::clusters::node::Node;
+use crate::api::FlightClient;
 use crate::configs::Config;
+use common_store_api_sdk::{StoreApiProvider, KVApi, ConnectionFactory};
+use kvlocal::LocalKVStore;
 
 pub type ClusterRef = Arc<Cluster>;
+pub type ClusterDiscoveryRef = Arc<ClusterDiscovery>;
 
-pub struct Cluster {
-    local_port: u16,
-    nodes: Mutex<HashMap<String, Arc<Node>>>,
+pub struct ClusterDiscovery {
+    local_id: String,
+    heartbeat: ClusterHeartbeat,
+    api_provider: Arc<Mutex<dyn NamespaceApi>>,
 }
 
-impl Cluster {
-    pub fn create_global(cfg: Config) -> Result<ClusterRef> {
-        Ok(Arc::new(Cluster {
-            nodes: Mutex::new(HashMap::new()),
-            local_port: Address::create(&cfg.query.flight_api_address)?.port(),
+impl ClusterDiscovery {
+    // TODO(Winter): this should be disabled by compile flag
+    async fn standalone_without_metastore(cfg: &Config) -> Result<ClusterDiscoveryRef> {
+        let local_id = global_unique_id();
+
+        let local_store = LocalKVStore::new_temp().await?;
```

Comment:
This api will change from `async` to `sync`.
I think all the meta api all `sync` is ok.
cc @dantengsky @zhang2014

Reply:
got it. by the way I think `async` is OK if it is not `&mut self`